### PR TITLE
runconfig: split resources into a struct

### PR DESCRIPTION
**Pull request by runcom**
Working on fixing test (it compiles btw)
Just opening this to see if it makes sense wrt, maybe, `docker set` on a container

/cc @tiborvass @vdemeester @unclejack @thaJeztah @coolljt0725 

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

File: runconfig/hostconfig.go [hunk 1]
```diff
@@ -181,40 +201,30 @@ type HostConfig struct {
...
+	ReadonlyRootfs  bool                  // Is the container root filesystem in read-only
+	SecurityOpt     []string              // List of string values to customize labels for MLS systems, such as SELinux.
+	UTSMode         UTSMode               // UTS namespace to use for the container
 
 	// Applicable to Windows
 	ConsoleSize [2]int         // Initial console size
 	Isolation   IsolationLevel // Isolation level of the container (eg default, hyperv)
+
+	// Contains container's resources (cgroups, ulimits)
+	*Resources
```
Comment (crosbymichael) [inline runconfig/hostconfig.go]: I would probably not have this as a pointer

Comment (runcom) [inline runconfig/hostconfig.go]: Me neither, I'll change it

Comment (runcom) [inline runconfig/hostconfig.go]: fixed this

**Comment by vdemeester** [comment]
@runcom some unit tests are failing :(

```bash
# github.com/docker/docker/daemon
daemon/daemon_unix_test.go:25: unknown runconfig.HostConfig field 'CPUShares' in struct literal
daemon/daemon_unix_test.go:63: unknown runconfig.HostConfig field 'CPUShares' in struct literal
```

**Comment by runcom** [comment]
@vdemeester on them

**Comment by calavera** [comment]
oh, man. `docker inspect` is not going to be happy about this. Is it really worth it?

**Comment by runcom** [comment]
if `inspect` needs to be really hacked I'd say no, but didn't tested inspect yet tbh..

**Comment by runcom** [comment]
@calavera can you tell me where inspect could break? isn't it time to have a `types` in `api` to expose things in `runconfig` also? :smiling_imp:

**Comment by vdemeester** [comment]
@calavera @runcom I think it's safe as `Resources` is *embedded*, right ?

**Comment by crosbymichael** [comment]
no issues with it being embeded for `docker inspect`

LGTM

**Comment by thaJeztah** [comment]
thanks @runcom !

**Comment by vdemeester** [comment]
LGTM for me too 🐲.

**Comment by thaJeztah** [comment]
moved to code review, but happy to hear if @calavera has objections

File: builder/dockerfile/internals.go [hunk 2]
```diff
@@ -498,18 +498,22 @@ func (b *Builder) create() (*daemon.Container, error) {
...
+	}
+
 	// TODO: why not embed a hostconfig in builder?
 	hostConfig := &runconfig.HostConfig{
-		CPUShares:    b.CPUShares,
-		CPUPeriod:    b.CPUPeriod,
-		CPUQuota:     b.CPUQuota,
-		CpusetCpus:   b.CPUSetCpus,
-		CpusetMems:   b.CPUSetMems,
 		CgroupParent: b.CgroupParent,
```
Comment (tiborvass) [inline builder/dockerfile/internals.go]: i wonder about cgrouparent

Comment (thaJeztah) [inline builder/dockerfile/internals.go]: wondering if it should moved as well (I'd be +1 for consistency), or if we should/can allow changing  it at runtime?

Comment (runcom) [inline builder/dockerfile/internals.go]: actually thought about cgroupParent, wasn't sure...

Comment (vdemeester) [inline builder/dockerfile/internals.go]: Hum what *use case* would it be to be able to *set* it ? :P

Comment (runcom) [inline builder/dockerfile/internals.go]: moving the container to a different cgroup which has some other constraints maybe, I'm moving it

**Comment by runcom** [comment]
@tiborvass @thaJeztah updated

**Comment by calavera** [comment]
@crosbymichael well, there are actually two issues:

1. it completely modifies the response api without keeping backwards compatibility.
2. it can break `--format` templates that reference those fields.

**Comment by runcom** [comment]
@calavera to 1. I can't see where this is modifying the response api, could you point me to where this is happening?
Before this PR:
```
{
    "Id": "e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963",
    "Created": "2015-11-20T18:37:06.307826142Z",
    "Path": "true",
    "Args": [],
    "State": {
        "Status": "exited",
        "Running": false,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 0,
        "ExitCode": 0,
        "Error": "",
        "StartedAt": "2015-11-20T18:37:06.472088819Z",
        "FinishedAt": "2015-11-20T18:37:06.500410287Z"
    },
    "Image": "c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7",
    "ResolvConfPath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/resolv.conf",
    "HostnamePath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/hostname",
    "HostsPath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/hosts",
    "LogPath": "",
    "Name": "/dreamy_bhaskara",
    "RestartCount": 0,
    "Driver": "overlay",
    "MountLabel": "",
    "ProcessLabel": "",
    "AppArmorProfile": "",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": null,
        "ContainerIDFile": "",
        "CpuShares": 0,
        "LogConfig": {
            "Type": "journald",
            "Config": {}
        },
        "NetworkMode": "default",
        "PortBindings": {},
        "RestartPolicy": {
            "Name": "no",
            "MaximumRetryCount": 0
        },
        "VolumeDriver": "",
        "VolumesFrom": null,
        "BlkioWeight": 0,
        "BlkioWeightDevice": null,
        "CapAdd": null,
        "CapDrop": null,
        "CgroupParent": "",
        "CpuPeriod": 0,
        "CpuQuota": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "Devices": [],
        "Dns": [],
        "DnsOptions": [],
        "DnsSearch": [],
        "ExtraHosts": null,
        "GroupAdd": null,
        "IpcMode": "",
        "KernelMemory": 0,
        "Links": null,
        "Memory": 0,
        "MemoryReservation": 0,
        "MemorySwap": 0,
        "MemorySwappiness": -1,
        "OomKillDisable": false,
        "PidMode": "",
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "SecurityOpt": null,
        "Ulimits": null,
        "UTSMode": "",
        "ShmSize": 67108864,
        "ConsoleSize": [
            0,
            0
        ],
        "Isolation": ""
    },
    "GraphDriver": {
        "Name": "overlay",
        "Data": {
            "LowerDir": "/var/lib/docker/0.0/overlay/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/root",
            "MergedDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/merged",
            "UpperDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/upper",
            "WorkDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/work"
        }
    },
    "Mounts": [],
    "Config": {
        "Hostname": "e26d88fe2ec5",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": true,
        "AttachStderr": true,
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": null,
        "Cmd": [
            "true"
        ],
        "Image": "busybox",
        "Volumes": null,
        "WorkingDir": "",
        "Entrypoint": null,
        "OnBuild": null,
        "Labels": {},
        "StopSignal": "SIGTERM"
    },
    "NetworkSettings": {
        "Bridge": "",
        "SandboxID": "",
        "HairpinMode": false,
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "Ports": null,
        "SandboxKey": "",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null,
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "MacAddress": "",
        "Networks": {
            "bridge": {
                "EndpointID": "",
                "Gateway": "",
                "IPAddress": "",
                "IPPrefixLen": 0,
                "IPv6Gateway": "",
                "GlobalIPv6Address": "",
                "GlobalIPv6PrefixLen": 0,
                "MacAddress": ""
            }
        }
    }
}
```
After:
```
{
    "Id": "e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963",
    "Created": "2015-11-20T18:37:06.307826142Z",
    "Path": "true",
    "Args": [],
    "State": {
        "Status": "exited",
        "Running": false,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 0,
        "ExitCode": 0,
        "Error": "",
        "StartedAt": "2015-11-20T18:37:06.472088819Z",
        "FinishedAt": "2015-11-20T18:37:06.500410287Z"
    },
    "Image": "c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7",
    "ResolvConfPath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/resolv.conf",
    "HostnamePath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/hostname",
    "HostsPath": "/var/lib/docker/0.0/containers/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/hosts",
    "LogPath": "",
    "Name": "/dreamy_bhaskara",
    "RestartCount": 0,
    "Driver": "overlay",
    "MountLabel": "",
    "ProcessLabel": "",
    "AppArmorProfile": "",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": null,
        "ContainerIDFile": "",
        "LogConfig": {
            "Type": "journald",
            "Config": {}
        },
        "NetworkMode": "default",
        "PortBindings": {},
        "RestartPolicy": {
            "Name": "no",
            "MaximumRetryCount": 0
        },
        "VolumeDriver": "",
        "VolumesFrom": null,
        "CapAdd": null,
        "CapDrop": null,
        "Dns": [],
        "DnsOptions": [],
        "DnsSearch": [],
        "ExtraHosts": null,
        "GroupAdd": null,
        "IpcMode": "",
        "Links": null,
        "OomKillDisable": false,
        "PidMode": "",
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "SecurityOpt": null,
        "UTSMode": "",
        "ShmSize": 67108864,
        "ConsoleSize": [
            0,
            0
        ],
        "Isolation": "",
        "CpuShares": 0,
        "CgroupParent": "",
        "BlkioWeight": 0,
        "BlkioWeightDevice": null,
        "CpuPeriod": 0,
        "CpuQuota": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "Devices": [],
        "KernelMemory": 0,
        "Memory": 0,
        "MemoryReservation": 0,
        "MemorySwap": 0,
        "MemorySwappiness": -1,
        "Ulimits": null
    },
    "GraphDriver": {
        "Name": "overlay",
        "Data": {
            "LowerDir": "/var/lib/docker/0.0/overlay/c51f86c283408d1749d066333f7acd5d33b053b003a61ff6a7b36819ddcbc7b7/root",
            "MergedDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/merged",
            "UpperDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/upper",
            "WorkDir": "/var/lib/docker/0.0/overlay/e26d88fe2ec513088e1be100222f954224f1a3ab45bbbd14a13e970463c03963/work"
        }
    },
    "Mounts": [],
    "Config": {
        "Hostname": "e26d88fe2ec5",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": true,
        "AttachStderr": true,
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": null,
        "Cmd": [
            "true"
        ],
        "Image": "busybox",
        "Volumes": null,
        "WorkingDir": "",
        "Entrypoint": null,
        "OnBuild": null,
        "Labels": {},
        "StopSignal": "SIGTERM"
    },
    "NetworkSettings": {
        "Bridge": "",
        "SandboxID": "",
        "HairpinMode": false,
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "Ports": null,
        "SandboxKey": "",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null,
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "MacAddress": "",
        "Networks": {
            "bridge": {
                "EndpointID": "",
                "Gateway": "",
                "IPAddress": "",
                "IPPrefixLen": 0,
                "IPv6Gateway": "",
                "GlobalIPv6Address": "",
                "GlobalIPv6PrefixLen": 0,
                "MacAddress": ""
            }
        }
    }
}
```

**Comment by runcom** [comment]
I'll test `--format` in a while

**Comment by crosbymichael** [comment]
@calavera its embeded, nothing changes in marshaled json.   Its no different from the example below

https://play.golang.org/p/uEEnyaKDkb

and yes, --format templates will need to be double checked but nothing changes or breaks the api response.

**Comment by runcom** [comment]
I'm testing `--format` and everything seems to be working 
```
docker inspect --format '{{.HostConfig.CpuShares}}' e26
0
```

**Comment by runcom** [comment]
The only difference in `--format` is I can query the embedded struct (in json as well):
```
docker inspect --format '{{.HostConfig.Resources}}' e26
{0  0 [] 0 0   [] 0 0 0 0 0xc8206cf170 []}
```

**Comment by calavera** [comment]
oh, the structure is embedded in hostConfig, nevermind, I thought it was a new field.

**Comment by calavera** [comment]
ok, LGTM. Waiting for the build to be green.

**Comment by runcom** [comment]
@calavera what about https://github.com/docker/docker/pull/18073#issuecomment-158491840

**Comment by calavera** [comment]
It's fine, it already happens with other embedded structs, we just don't document that behavior.

**Comment by runcom** [comment]
cool

**Comment by cpuguy83** [comment]
LGTM

**Comment by vdemeester** [comment]
Windows failure unrelated :sweat_smile:

**Pull request merged**